src/Data/Nat/Binary/Properties.agda

```diff
@@ -273,11 +274,35 @@ private
 <-trans : Transitive _<_
 <-trans = <-Monomorphism.trans ℕₚ.<-trans
 
-<-cmp :  ∀ (x y) → Tri (x < y) (x ≡ y) (x > y)
-<-cmp = <-Monomorphism.compare ℕₚ.<-cmp
+-- Comparisons and decidability are not implemented via the morphism so as
+-- to avoid reverting to linear time for the decision procedures.
+<-cmp : Trichotomous _≡_ _<_
+<-cmp zero     zero      = tri≈ x≮0    refl  x≮0
+<-cmp zero     2[1+ _ ]  = tri< 0<even (λ()) x≮0
+<-cmp zero     1+[2 _ ]  = tri< 0<odd  (λ()) x≮0
+<-cmp 2[1+ _ ] zero      = tri> (λ())  (λ()) 0<even
+<-cmp 2[1+ x ] 2[1+ y ]  with <-cmp x y
+... | tri< x<y _    _   =  tri< lt (<⇒≢ lt) (<⇒≯ lt)  where lt = even<even x<y
+... | tri≈ _   refl _   =  tri≈ (<-irrefl refl) refl (<-irrefl refl)
+... | tri> _   _    x>y =  tri> (>⇒≮ gt) (>⇒≢ gt) gt  where gt = even<even x>y
+<-cmp 2[1+ x ] 1+[2 y ]  with <-cmp x y
+... | tri< x<y _    _ =  tri< lt (<⇒≢ lt) (<⇒≯ lt)  where lt = even<odd x<y
+... | tri≈ _   refl _ =  tri> (>⇒≮ gt) (>⇒≢ gt) gt
+  where
+  gt = subst (_< 2[1+ x ]) refl (1+[2x]<2[1+x] x)
+... | tri> _ _ y<x =  tri> (>⇒≮ gt) (>⇒≢ gt) gt  where gt = odd<even (inj₁ y<x)
+<-cmp 1+[2 _ ] zero =  tri> (>⇒≮ gt) (>⇒≢ gt) gt  where gt = 0<odd
+<-cmp 1+[2 x ] 2[1+ y ]  with <-cmp x y
+... | tri< x<y _ _ =  tri< lt (<⇒≢ lt) (<⇒≯ lt)  where lt = odd<even (inj₁ x<y)
+... | tri≈ _ x≡y _ =  tri< lt (<⇒≢ lt) (<⇒≯ lt)  where lt = odd<even (inj₂ x≡y)
+... | tri> _ _ x>y =  tri> (>⇒≮ gt) (>⇒≢ gt) gt  where gt = even<odd x>y
+<-cmp 1+[2 x ] 1+[2 y ]  with <-cmp x y
+... | tri< x<y _  _   =  tri< lt (<⇒≢ lt) (<⇒≯ lt)  where lt = odd<odd x<y
+... | tri≈ _ refl _   =  tri≈ (≡⇒≮ refl) refl (≡⇒≯ refl)
+... | tri> _ _    x>y =  tri> (>⇒≮ gt) (>⇒≢ gt) gt  where gt = odd<odd x>y
 
```

Comment:
Feels like we should be able to play @laMudri's games on `Tri`
just like we have on `Dec`.

Reply:
Yup, probably true, although won't we need need a three valued logic? Might be best to open an issue for `v1.3`...

Reply:
At least in this case, we have a clear intended semantics for our three-valued logic (if we choose to see it as such). But yeah, the same pattern clearly applies.

Reply:
We can use the classic `LT`, `EQ`, `GT` familiar to Haskellers.